highlighters/main/main-highlighter.zsh

```diff
@@ -531,62 +544,46 @@ _zsh_highlight_main_highlighter_highlight_list()
       # Expand aliases.
       _zsh_highlight_main__type "$arg"
       local res="$REPLY"
-      if [[ $res == "alias" ]]; then
-        () {
-        local -A seen_alias
-        while [[ $REPLY == alias ]]; do
+      if [[ $res == "alias" ]] && [[ $last_alias != $arg ]]; then
+        # Avoid looping forever on alias a=b b=c c=b, but allow alias foo='foo bar'
+        if (( $+seen_alias[$arg] )); then
+          _zsh_highlight_main_add_region_highlight $start_pos $end_pos unknown-token
+          continue
+        fi
         seen_alias[$arg]=1
+        last_alias=$arg
         _zsh_highlight_main__resolve_alias $arg
-        # Use a temporary array to ensure the subscript is interpreted as
-        # an array subscript, not as a scalar subscript
         local -a alias_args
-        # TODO: the ${interactive_comments+set} path needs to skip comments; see test-data/alias-comment1.zsh
         alias_args=( ${interactive_comments-${(z)REPLY}}
                      ${interactive_comments+${(zZ+c+)REPLY}} )
-        # Avoid looping forever on alias a=b b=c c=b, but allow alias foo='foo bar'
-        [[ $arg == $alias_args[1] ]] && break
-        arg=$alias_args[1]
-        if (( $+seen_alias[$arg] )); then
-          res=none
-          break
-        fi
-        _zsh_highlight_main__type "$arg"
-        done
-        }
-        _zsh_highlight_main_highlighter_expand_path $arg
-        arg=$REPLY
-        () {
-        # Make sure to use $arg_raw here, rather than $arg.
-        integer insane_alias
-        case $arg_raw in
+        case $arg in
           # Issue #263: aliases with '=' on their LHS.
           #
           # There are three cases:
           #
           # - Unsupported, breaks 'alias -L' output, but invokable:
           ('='*) :;;
           # - Unsupported, not invokable:
-          (*'='*) insane_alias=1;;
+          (*'='*)
+            _zsh_highlight_main_add_region_highlight $start_pos $end_pos unknown-token
+            continue
+            ;;
           # - The common case:
           (*) :;;
         esac
-        if (( insane_alias )); then
-          style=unknown-token
-        # Calling 'type' again; since __type memoizes the answer, this call is just a hash lookup.
-        elif ! _zsh_highlight_main__type "$arg" || [[ $REPLY == 'none' ]]; then
-          style=unknown-token
+        args=( $alias_args $args )
+        if (( in_alias == 0 )); then
+          _zsh_highlight_main_add_region_highlight $start_pos $end_pos alias
+          (( in_alias += $#alias_args + 1 )) # Add one because in_alias counts $arg as well
```

Comment:
If I understand correctly, the ` + 1` is because the _current_ iteration handles the alias, and the next `$#alias_args` iterations will handle its expansion, is that correct?  I think it might be clearer if it were said in more words?

Reply:
The +1 is because we will `in_alias--` on the next loop iteration so this iteration should be considered in `in_alias` as well. In the case that we're already `in_alias` this arg is already included in the count. It makes sense in my brain; hopefully that helps?